src/main/java/hudson/plugins/build_timeout/BuildStepWithTimeout.java

```diff
@@ -0,0 +1,153 @@
+            this.operationList = operationList;
+        }
+        else {
+            this.operationList = Collections.emptyList();
+        }
+
+    }
+
+    public List<BuildTimeOutOperation> getOperationList() {
+        return operationList;
+    }
+
+
+    public BuildStep getBuildStep() {
+        return buildStep;
+    }
+
+    public void setBuildStep(BuildStep buildStep) {
+        this.buildStep = buildStep;
+    }
+
+    private long getTimeout(Run run, TaskListener listener) throws IOException, InterruptedException {
+        try {
+            return strategy.getTimeOut((AbstractBuild<?, ?>) run, (BuildListener) listener);
+        } catch (MacroEvaluationException e) {
+            e.printStackTrace(listener.getLogger());
+            listener.error("Can't evaluate timeout - timeout would be disabled");
+            return Long.MAX_VALUE;
+        }
+    }
+
+
+    public boolean perform(final Build<?,?> build, final Launcher launcher, final BuildListener listener) throws InterruptedException, IOException {
+        final Timer timer = new Timer("Timer-" + build.getDisplayName().replace(" ", "_"), true);
+        final AtomicBoolean stopped = new AtomicBoolean(false);
+        try {
+            final long delay = getTimeout(build, listener);
+
+            timer.schedule(new TimerTask() {
+                @Override
+                public void run() {
+                    if (operationList.isEmpty()) {
+                        new AbortOperation().perform(build, listener, delay);
+                    }
+
+                    for (BuildTimeOutOperation op : operationList) {
+                        op.perform(build, listener, delay);
+                    }
+                    stopped.set(true);
+                    timer.cancel();
+                }
+            }, delay);
+
+            return buildStep.perform((AbstractBuild) build, launcher, listener);
+        } catch(InterruptedException e) {
+            if (!stopped.get()) {
+                throw e;
+            }
+        }
+
+        return false;
+    }
+
+    public BuildTimeOutStrategy getStrategy() {
+        return strategy;
+    }
+
+    @Override
+    public BuildStepMonitor getRequiredMonitorService() {
+        return BuildStepMonitor.NONE;
+    }
+
+    @Extension
+    public static class DescriptorImpl extends BuildStepDescriptor<Builder> {
+        public DescriptorImpl() {
+            load();
+        }
+
+        @Override
+        public BuildStepWithTimeout newInstance(StaplerRequest req, JSONObject formData)
+                throws hudson.model.Descriptor.FormException {
+            BuildStep buildstep = BuildTimeOutUtility.bindJSONWithDescriptor(req, formData, "buildStep", BuildStep.class);
+            BuildTimeOutStrategy strategy = BuildTimeOutUtility.bindJSONWithDescriptor(req, formData, "strategy", BuildTimeOutStrategy.class);
+
+            if (!(strategy instanceof AbsoluteTimeOutStrategy || strategy instanceof DeadlineTimeOutStrategy)) {
```

Comment:
Add a new method like `TimeOutStrategy#isApplicableAsBuildStep` and filter with that.
That's not a good idea to list up applicable classes in codes.